### PR TITLE
feat: Implement graceful shutdown in application-controller

### PR DESCRIPTION
**Pull request by leehosu**
### summary

Add logic to verify Gracefulshutdown.

### before

스크린샷 2024-08-10 오후 4 23 40

### after
스크린샷 2024-08-10 오후 4 36 07

related #18642 

Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [ ] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [ ] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [ ] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [ ] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [ ] Optional. My organization is added to USERS.md.
* [ ] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

### CODE REVIEW

File: cmd/argocd-application-controller/commands/argocd_application_controller.go [hunk 1]
```diff
@@ -188,10 +192,26 @@ func NewCommand() *cobra.Command {
...
+				cancel()
+				wg.Done()
+			}()
+
+			go appController.Run(ctx, statusProcessors, operationProcessors, &wg)
 
 			// Wait forever
-			select {}
+			// select {}
+			wg.Wait()
```
Comment (crenshaw-dev) [inline cmd/argocd-application-controller/commands/argocd_application_controller.go:211]: Was testing a bit locally, and I think we can get the same effect by just replacing `select {}` with `<-ctx.Done()`. At least the behavior seems the same. What do you think?

**Comment by crenshaw-dev** [comment]
Good indication that the change allowed coverage data to be produced:

image

File: cmd/argocd-application-controller/commands/argocd_application_controller.go (hunk 1 above)
Comment (leehosu) [reply cmd/argocd-application-controller/commands/argocd_application_controller.go:211]: Wg.Done() looks better because it can be an infinite atmosphere with the existing select{}. And above all, if you consider unity elsewhere, wg.Done() looks better. What do you think?

Comment (crenshaw-dev) [reply cmd/argocd-application-controller/commands/argocd_application_controller.go:211]: > it can be an infinite atmosphere with the existing select{}

Can you rephrase? I'm not sure what you mean "infinite atmosphere." I'm suggesting removing `select {}` and instead doing `<-ctx.Done()`.

> if you consider unity elsewhere, wg.Done() looks better

If you're looking at the repo-server graceful shutdown code... I wrote that code, and now I don't like it. 😆 Looks like it can be simplified.

Comment (leehosu) [reply cmd/argocd-application-controller/commands/argocd_application_controller.go]: Deleted select {}. The controller you're working on now doesn't seem to have a problem.

I'll think about repo-server in the future as well. And If possible, I'll also proceed with refactoring on the repo-server side.

Comment (crenshaw-dev) [reply cmd/argocd-application-controller/commands/argocd_application_controller.go]: Can you take a look at this? I think we can simplify further. https://github.com/leehosu/argo-cd/pull/1

Comment (leehosu) [reply cmd/argocd-application-controller/commands/argocd_application_controller.go]: So Cool. That's simple. I checked and merged into the branch.

**Review by crenshaw-dev — APPROVED** [review]
Thanks, @leehosu!
